### PR TITLE
New funtions 2018

### PR DESCRIPTION
**Pull request by pawelkisielewicz**

### CODE REVIEW

File: employees/views.py [hunk 1]
```diff
@@ -140,94 +154,118 @@ def order_by_unpaid_salaries(name_filter, descending):
...
+
+    def post(self, request):
+
+        if not request.is_ajax() or not request.method == 'POST':
+            from django.http import HttpResponseNotAllowed
+            return HttpResponseNotAllowed(['POST'])
+        try:
+            order = self.request.POST['order']
+            request.session['order'] = order
+        except:
```
Comment (pawelkisielewicz) [inline employees/views.py]: do poprawy try except

File: employees/views.py [hunk 2]
```diff
@@ -140,94 +154,118 @@ def order_by_unpaid_salaries(name_filter, descending):
...
+            request.session['order'] = 'last_name'
```
Comment (pawelkisielewicz) [inline employees/views.py]: correct try except

File: employees/views_business_logic.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+
+        old_date = employee_to_update.contract_exp_date
+        self.exp_date=old_date
+
+        try:
+            day = old_date.day
+            old_month = old_date.month
+            old_year = old_date.year
+        except:
```
Comment (Jakub89) [inline employees/views_business_logic.py]: Another nasty try/except block. We never should be force to use `try:` on assigning to variables.

File: employees/views_business_logic.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    exp_date=None
+
+    def add_one_month(self, employee_id):
+        employee_to_update = Employee.objects.get(pk=employee_id)
+
+        self.name=employee_to_update.first_name+ ' '+employee_to_update.last_name
+
+
+        old_date = employee_to_update.contract_exp_date
+        self.exp_date=old_date
```
Comment (Jakub89) [inline employees/views_business_logic.py]: Please stick to one styling. There should be space between "=" sign unless it's a function call like `.get(pk=employee_id)`.

File: employees/models.py [hunk 5]
```diff
@@ -201,21 +228,29 @@ class Month(models.Model):
...
-    salary_is_paid = models.BooleanField(_(u'Paid?'),default=False, choices=bool_choices)
-    hours_worked_in_this_month = models.DecimalField(_('hours worked in this month'),decimal_places=1, max_digits=4, default=0,
-                                                    validators=[MaxValueValidator(720), MinValueValidator(0)])
-    month_is_approved = models.BooleanField(_(u'Approved?'),default=False, choices=bool_choices)
-    rate_per_hour_this_month = models.DecimalField(_('rate per hour this month'),decimal_places=2, max_digits=7, default=0,
+    month = models.IntegerField(_(u'Month'), choices=month_choices, default=default_month)
+    salary_is_paid = models.BooleanField(_(u'Paid?'), default=False, choices=bool_choices)
+    hours_worked_in_this_month = models.DecimalField(_('hours worked in this month'), decimal_places=1, max_digits=4,
+                                                     default=0,
+                                                     validators=[MaxValueValidator(720), MinValueValidator(0)])
```
Comment (Jakub89) [inline employees/models.py]: This is unreadable fix it like this:
```
hours_worked_in_this_month = models.DecimalField(
    _('hours worked in this month'), 
    decimal_places=1, 
    max_digits=4,
    default=0,
    validators=[MaxValueValidator(720), MinValueValidator(0)]
)
```
And stick to similar writhing in others parts of the code ;)

File: employees/functional_tests/employees_tests.py [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+            element = None
+        self.assertNotEqual(element, None, 'unable to find next page label with status clickable')
+
+    def __check_page_exist_and_get_true_if_selected(self, number):
+        try:
+            id = 'pg-label' + str(number) + '-checked'
+            self.selenium.find_element_by_id(id)  # test if element exist. If doesn't throws exceptions
+            return True
+        except:
+            try:
```
Comment (Jakub89) [inline employees/functional_tests/employees_tests.py]: This is strait abomination, the whole test need to be rethink and rewritten. Use try/except as little as possible. And never one inside another.

**Review by Jakub89 — CHANGES_REQUESTED** [review]

File: employees/views.py [hunk 7]
```diff
@@ -140,94 +154,118 @@ def order_by_unpaid_salaries(name_filter, descending):
...
+        try:
+            per_page = self.request.POST['per_page']
+            request.session['per_page'] = per_page
+        except:
+            request.session['per_page'] = 10
+
+        try:
+            page = self.request.POST['page']
+            request.session['page'] = page
+        except:
```
Comment (Jakub89) [inline employees/views.py]: to many try/except blocks

File: employees/views.py [hunk 8]
```diff
@@ -140,94 +154,118 @@ def order_by_unpaid_salaries(name_filter, descending):
...
+
+    @staticmethod
+    def __make_true_or_false_from_POST_request(request, filer_name):
+        try:
+            value = request.POST[filer_name]
+            if value == 'on':
+                return True
+            else:
+                return False
+        except:
```
Comment (Jakub89) [inline employees/views.py]: Same here and to much logic inside `try:` block. You should limit yourself to single line where You think exception could appear.

File: employees/tests.py [hunk 9]
```diff
@@ -288,142 +290,302 @@ def test_should_month_update_view_work(self):
...
+
+            self.assertEqual(contractExtension.add_three_months(key), extend_status)
+            exp_date = dict.get(key)[1]
+
+            self.assertEqual(contractExtension.exp_date, exp_date)
+
+
+# class ContractExtensionViewStatusTests(TestCase):
+#     fixtures = ['users_myuser.json', 'employees_employee.json', 'month.json']
+#
```
Comment (Jakub89) [inline employees/tests.py]: Delete all commented code.

File: employees/tests.py [hunk 10]
```diff
@@ -288,142 +290,302 @@ def test_should_month_update_view_work(self):
...
+#         self.assertEqual(status, answer_status)
+
+
+class ContractExtensionViewCorrectDateTests(TestCase):
+    fixtures = ['users_myuser.json', 'employees_employee.json', 'month.json']
+
+    # fixtures = ['fikstura_cities','users_myuser.json', 'employees_employee.json']
+    # cities deleted manual from fixtury. To test with cities activate upper line and use employees_employee_with_cities.json
+    # reason- loading fikstura_cities takes very long time
```
Comment (Jakub89) [inline employees/tests.py]: same here

File: employees/tests.py [hunk 11]
```diff
@@ -288,142 +290,302 @@ def test_should_month_update_view_work(self):
...
+                6: [True, '2018-01-19', 'Andrzej Strzelba'],
+                12: [True, '2017-03-02', 'Ccc Ccc'],
+                11: [False, 'null', 'Bbb Bbb'],
+                29: [False, 'null', 'Paweł Testowy'],
+                }
+
+        for key in dict.keys():
+            self.__should_contract_extension_work_properly(key, dict.get(key)[0], dict.get(key)[1], dict.get(key)[2],
+                                                           'add_1_id')
+            # self.__should_contract_extension_work_properly(key, dict.get(key), 'add_3_month')
```
Comment (Jakub89) [inline employees/tests.py]: same here and fix style in function above.

File: employees/tests.py [hunk 12]
```diff
@@ -288,142 +290,302 @@ def test_should_month_update_view_work(self):
...
 
-    def test_should_hide_paid_employees_filter_work_properly(self):
-        admin_login(self)
-        url = reverse_lazy('employees')
-        unpaid_response = self.c.get(url + "?hide_paid_employees_filter=on")
-        self.assertQuerysetEqual(
-            unpaid_response.context['all_employee_list'],
-            ['<Employee: Czarny Lodziarz>']
-        )
+        import json
```
Comment (Jakub89) [inline employees/tests.py]: all `import` should be at the beginning of file.

File: employees/tests.py [hunk 13]
```diff
@@ -288,142 +290,302 @@ def test_should_month_update_view_work(self):
...
+        date = dict[2]
+
+        self.assertEqual(status, answer_status)
+        self.assertEqual(name, expected_name)
+        self.assertIn(expected_date, date)
 
-class MonthFilterTests(TestCase):
 
+@skip
+class MonthFilterTests(TestCase):
```
Comment (Jakub89) [inline employees/tests.py]: Why is it skipped? Can it be fixed? If not maybe the whole test should be dropped.

File: employees/tests.py [hunk 14]
```diff
@@ -448,152 +610,147 @@ def test_should_year_filters_work(self):
...
         self.assertQuerysetEqual(response.context['months'],
-      
-                           ['<Month: January 2018>'])
 
-class MonthApproveTests(TestCase):
+                                 ['<Month: January 2018>'])
+
 
+@skip
+class MonthApproveTests(TestCase):
```
Comment (Jakub89) [inline employees/tests.py]: same here

File: employees/tests.py [hunk 15]
```diff
@@ -448,152 +610,147 @@ def test_should_year_filters_work(self):
...
         month.refresh_from_db()
-        
-        self.assertEqual(month.month_is_approved, True)        
 
-class CityAutocompleteTests(TestCase):
+        self.assertEqual(month.month_is_approved, True)
+
 
+@skip
+class CityAutocompleteTests(TestCase):
```
Comment (Jakub89) [inline employees/tests.py]: and here

File: employees/tests_pyTest.py [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+
+
+#
+# from employees import views
+#
+
+
+# class EmployeeDetailViewTests(object):
+# from employees.views import EmployeeAction
+# from employees.views_business_logic import EmployeeAction
```
Comment (Jakub89) [inline employees/tests_pyTest.py]: Them commentzzz

File: employees/tests_pyTest.py [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+class TestEmployeeAction(object):
+    fixtures = ['users_myuser.json', 'employees_employee.json', 'month.json']
+
+    # def test_one(self):
+    #     x = "this"
+    #     assert 'h' in x
+    #
+    # def test_two(self):
+    #     x = "hello"
+    #     assert hasattr(x, 'check')
```
Comment (Jakub89) [inline employees/tests_pyTest.py]: again

File: employees/tests_pyTest.py [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+# @pytest.mark.parametrize("message", ["nie zgadzam sie"])
+# def test_employee_message_view_with_the_existing_record_param(id, message):
+#     pass
+# # def test_employee_message_view_with_the_existing_record_param(self, id, message):
+# #     url = reverse_lazy('employee_message', args=(id,))
+# #     response = self.client.get(url, follow=True)
+# #     self.assertEqual(response.status_code, 200)
+# #     self.assertContains(response, message)
+# #     # assert eval(message in response)
+#
```
Comment (Jakub89) [inline employees/tests_pyTest.py]: and again

File: employees/views.py [hunk 19]
```diff
@@ -83,39 +91,45 @@ def make_year_list_for_filtering_in_emp_detail(employee):
...
-            queryset = queryset.exclude(position=positions[key])
-    queryset = queryset.exclude(position=None)
-    return queryset
+
+def former_and_current_filtering(self, queryset):
+    # queryset = queryset.exclude(currently_employed=None) don't need to take caro of None, column Not Null in Data Base
+    if self.request.is_ajax():
+        if self.request.GET.get('current_employees'):
+            queryset = queryset.exclude(currently_employed=False)
+        if not self.request.GET.get('current_employees'):
```
Comment (Jakub89) [inline employees/views.py]: This should be replaced with `else` statement since there are only two options.

File: employees/views.py [hunk 20]
```diff
@@ -83,39 +91,45 @@ def make_year_list_for_filtering_in_emp_detail(employee):
...
+            queryset = queryset.exclude(currently_employed=True)
+
+        return queryset
+    else:
+        if self.request.session['current_employees']:
+            queryset = queryset.exclude(currently_employed=False)
+        if not self.request.session['current_employees']:
```
Comment (Jakub89) [inline employees/views.py]: same here

File: employees/views.py [hunk 21]
```diff
@@ -140,94 +154,118 @@ def order_by_unpaid_salaries(name_filter, descending):
...
         if self.request.is_ajax():
-            qset = context['all_employee_list']
-            order = context['orderby']
             try:
-                page_obj = context['page_obj']
-                paginator = context['paginator']
+                value = self.request.GET.get(filer_name)
+                if value == 'on':
+                    return True
+                if value == True:
```
Comment (Jakub89) [inline employees/views.py]: More readable would be:
```
if value == 'on' or value is True:
    return True
```

**Review by Jakub89 — CHANGES_REQUESTED** [review]

File: employees/views_business_logic.py (hunk 3 above)
Comment (pawelkisielewicz) [reply employees/views_business_logic.py]: done.. good to have tests for this class, so i know it still works properly  :)

File: employees/views_business_logic.py (hunk 4 above)
Comment (pawelkisielewicz) [reply employees/views_business_logic.py]: done for the whole file

File: employees/functional_tests/employees_tests.py (hunk 6 above)
Comment (pawelkisielewicz) [reply employees/functional_tests/employees_tests.py]: is it ok for simple method checking if element exist. I want to add custom comment to exception:

    def __check_if_previous_page_exist_and_is_clickable(self):
        try:
            self.selenium.find_element_by_xpath('//label[@class="my-button previous clickable"]')
        except NoSuchElementException:
            raise NoSuchElementException('unable to find previous page label with status clickable')

File: employees/views.py (hunk 7 above)
Comment (pawelkisielewicz) [reply employees/views.py]: yes... fixed :)

File: employees/tests.py (hunk 10 above)
Comment (pawelkisielewicz) [reply employees/tests.py]: look again, its not dead code. I start tests without cities with edited file employees (deleted cities). It works good and i left message why i did it. Fixed to "nice" comment. Is it OK?

File: employees/tests.py (hunk 12 above)
Comment (pawelkisielewicz) [reply employees/tests.py]: done

File: employees/tests_pyTest.py (hunk 18 above)
Comment (pawelkisielewicz) [reply employees/tests_pyTest.py]: this whole class and file was unfinished. I wrote new test, not a perfect one, but it was hard to find function that could be tested with pytest and parameters.

File: employees/views.py (hunk 19 above)
Comment (pawelkisielewicz) [reply employees/views.py]: fixed :)

File: employees/tests_pyTest.py (hunk 18 above)
Comment (Jakub89) [reply employees/tests_pyTest.py]: No need for pytest, standard django tests should do fine for now.

File: employees/tests.py (hunk 13 above)
Comment (pawelkisielewicz) [reply employees/tests.py]: fixed, it was your old tests. Some are rewritten now, redundant deleted

File: employees/tests_pyTest.py (hunk 18 above)
Comment (pawelkisielewicz) [reply employees/tests_pyTest.py]: Right. But this project is for me to learn and i wanted to write at least one test like this to learn it :)

**Pull request merged**